tcms/testplans/tests/test_views.py

```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
+        self.assertTrue(plan.emailing.notify_on_case_update)
+
+    def test_get_with_no_perm_redirects_to_login(self):
+        remove_perm_from_user(self.user, self.add_testplan_permission)
+        response = self.client.get(self.location, follow=True)
+        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+
+    def test_post_with_no_perm_redirects_to_login(self):
+        remove_perm_from_user(self.user, self.add_testplan_permission)
+        response = self.client.post(self.location, self.request, follow=True)
+        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+
+
+class EditPlanViewTest(BasePlanTest):
+    test_plan_name = 'plan name'
+
+    def setUp(self):
+        super().setUp()
+        self.testplan_edit_url = reverse('plan-edit', args=[self.plan_id])
+
+        user_should_have_perm(self.user, perm='testplans.change_testplan')
+        self.user.is_superuser = False
+        self.user.save()
+
+        self.testplan_edit_data = {
+            'author': self.user.pk,
+            'product': self.product.pk,
+            'product_version': self.product_version.pk,
+            'type': self.plan_type.pk,
+            'name': self.test_plan_name,
+
+            'email_settings-0-auto_to_plan_author': 'on',
+            'email_settings-0-auto_to_case_owner': 'on',
+            'email_settings-0-auto_to_case_default_tester': 'on',
+            'email_settings-0-notify_on_case_update': 'on',
+            'email_settings-0-notify_on_plan_update': 'on',
+
+            'email_settings-0-plan': '',
+            'email_settings-0-id': self.test_plan.emailing.pk,
+            'email_settings-TOTAL_FORMS': '1',
+            'email_settings-INITIAL_FORMS': '1',
+            'email_settings-MIN_NUM_FORMS': '0',
+            'email_settings-MAX_NUM_FORMS': '1',
+            'is_active': True,
+        }
+
+    def test_show_testplan_edit_page(self):
+        response = self.client.get(self.testplan_edit_url)
+        self.assertContains(response, _('Edit TestPlan'))
+        assert_notify_buttons(self, response)
+
+    def test_edit_testplan_text_field(self):
+        edit_data = self.testplan_edit_data.copy()
+        new_text = 'Edited: {0}'.format(self.test_plan.text)
+        edit_data['text'] = new_text
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data, follow=True)
+
+        self.assertContains(response, new_text)
+        self.test_plan.refresh_from_db()
+        self.assertEqual(new_text, self.test_plan.text)
+
+    def test_try_editting_testplan_with_invalid_form(self):
+        test_plan = TestPlanFactory(
+            author=self.user,
+            product=self.product,
+            product_version=self.product_version)
+
+        testplan_edit_url = reverse('plan-edit', args=[test_plan.pk])
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['email_settings-0-auto_to_case_owner'] = 'invalid_value'
```

Comment:
Hmmm, I'm not sure how that works in practices. We've never tried setting these fields to something different that on/off. In any case if you want to use this there must be an assert that this button shows an error message.

Another possibility is to set the `type` field to an invalid ID (e.g. -1) and there should be an error message - you can modify the value for the select field in the browser to play with it. Needs assert that there is an error message as well.

Reply:
Changing type to an invalid id does not cover here:

https://github.com/kiwitcms/Kiwi/blob/508564d0fd90bcdab725684c2804126f5092f0b0/tcms/testplans/views.py#L82

Since the validation here made for `notify_formset`. So i would suggest to keep it invalid over `email_settings-0-auto_to_case_owner`.

For error message check an additional case would be possible.

Reply:
Ah right, you want to cover the branch on line 83 and I've been thinking about the `or` expression in `get_context_data()` which is completely different thing.  Since there are 2 forms on this page - the notify one and the main one choose better names for the test methods and we won't get confused.